chipflow_lib/steps/silicon.py

```diff
@@ -4,29 +4,91 @@
 import sys
 import time
 import json
-import pprint
 import inspect
 import argparse
 import subprocess
 import importlib.metadata
 
 import requests
+from amaranth import *
+from amaranth.lib import io
+from amaranth.lib.cdc import FFSynchronizer
 
 from .. import ChipFlowError
-from ..platforms.silicon import SiliconPlatform
+from ..platforms import SiliconPlatform, top_interfaces
+
+class SiliconTop(Elaboratable):
+    def __init__(self, config={}):
+        self._config = config
+
+    def elaborate(self, platform: SiliconPlatform):
+        m = Module()
+
+        platform.instantiate_ports()
+
+        for clock, pin in self._clocks.items():
+            if clock == 'default':
+                clock = 'sync'
+            setattr(m.domains, clock,  ClockDomain(name=clock))
+            clk_buffer = io.Buffer("i", platform.request(pin))
+            setattr(m.submodules, "clk_buffer_" + clock, clk_buffer)
+            m.d.comb += ClockSignal().eq(clk_buffer.i)
+
+        rst_buffer = io.Buffer("i", ~platform.request(self._reset))
+        m.submodules.rst_buffer = rst_buffer
+        m.submodules.rst_sync = FFSynchronizer(rst_buffer.i, ResetSignal())
+
+        def _connect_interface(interface, name):
+            pins = dict()
+            for member in interface.signature.members:
+                pin, suffix = member.rsplit("_", 2)
+                assert suffix in ("o", "i", "oe"), suffix
+                pins[pin] = getattr(interface, member).width
+            for pin, width in pins.items():
+                for i in range(width):
+                    platform_pin = platform.request(f"{name}_{pin}{'' if width == 1 else str(i)}")
+                    if hasattr(interface, f"{pin}_i"):
+                        m.d.comb += getattr(interface, f"{pin}_i")[i].eq(platform_pin.i)
+                    if hasattr(interface, f"{pin}_o"):
+                        m.d.comb += platform_pin.o.eq(getattr(interface, f"{pin}_o")[i])
+                    if hasattr(interface, f"{pin}_oe"):
+                        m.d.comb += platform_pin.oe.eq(getattr(interface, f"{pin}_oe")[i])
+
+        # Clock generation
+        m.domains.sync = ClockDomain()
+
+        clk = platform.request("sys_clk")
+        m.d.comb += ClockSignal().eq(clk.i)
+        m.submodules.rst_sync = FFSynchronizer(
+            ~platform.request("sys_rst_n").i,
+            ResetSignal())
+
+        # heartbeat led (to confirm clock/reset alive)
+        heartbeat_ctr = Signal(23)
```

Comment:
maybe this should be optional?

Reply:
I'm just debating, could have `HeartBeat` as a top level component to add (in `[chipflow.top]`), or maybe better to have a `[chipflow.silicon.debug]` section where things like debug, test patterns, on chip test etc can be configured.

What do you think @gatecat?

Reply:
I think the debug option is probably the best, as that's where we could configure e.g. extra connections from the scan chain JTAG to user-specific debug logic, too.